frontend/src/components/teacher/results/StudentAnswersSection/CorrectedQuestion.tsx

```diff
@@ -0,0 +1,145 @@
+import React, { useMemo } from "react";
+import { Box } from "@chakra-ui/react";
+
+import type { QuestionComponentResponse } from "../../../../APIClients/types/TestClientTypes";
+import type {
+  ImageMetadata,
+  MultipleChoiceMetadata,
+  MultiSelectMetadata,
+  QuestionTextMetadata,
+  ShortAnswerMetadata,
+  TextMetadata,
+} from "../../../../types/QuestionMetadataTypes";
+import {
+  QuestionElementType,
+  ResponseElementType,
+} from "../../../../types/QuestionTypes";
+import Image from "../../../student/questions/question-elements/Image";
+import QuestionText from "../../../student/questions/question-elements/QuestionText";
+import Text from "../../../student/questions/question-elements/Text";
+
+import CorrectedMultipleChoice from "./question-elements/CorrectedMultipleChoice";
+import CorrectedMultiSelect from "./question-elements/CorrectedMultiSelect";
+import CorrectedShortAnswer from "./question-elements/CorrectedShortAnswer";
+
+interface QuestionProps {
+  elements: QuestionComponentResponse[];
+  studentAnswers: number[][];
+}
+
+const CorrectedQuestion = ({
+  elements,
+  studentAnswers,
+}: QuestionProps): React.ReactElement => {
+  const answerElementIndex = useMemo(() => {
+    let answersSoFar = 0;
+    return elements.map((element) => {
+      if (element.type in ResponseElementType) {
+        const elementIndex = answersSoFar;
+        answersSoFar += 1;
+        return elementIndex;
+      }
+      return 0;
+    });
+  }, [elements]);
+
+  const subquestionIndex = useMemo(() => {
+    let subquestionsSoFar = 0;
+    return elements.map((element) => {
+      if (element.type === QuestionElementType.QUESTION_TEXT) {
+        const elementIndex = subquestionsSoFar;
+        subquestionsSoFar += 1;
+        return elementIndex;
+      }
+      return 0;
+    });
+  }, [elements]);
```

Comment:
I _think_ we've repeated this logic a couple of times, if so, is it possible to refactor to a common function?